src/routers/v1/checkout/index.ts

```diff
@@ -3,14 +3,102 @@ import prisma from "@mirlo/prisma";
 
 import stripe from "../../../utils/stripe";
 
+const DEFAULT_CLIENT_BASE_URL =
+  process.env.REACT_APP_CLIENT_DOMAIN ?? "https://mirlo.space";
+
+const ensureTrailingSlash = (value: string) =>
+  value.endsWith("/") ? value : `${value}/`;
+
+const normaliseBaseUrl = (candidate?: string | null) => {
+  const fallback = ensureTrailingSlash(DEFAULT_CLIENT_BASE_URL);
+  if (!candidate) {
+    return fallback;
+  }
+
+  try {
+    const trimmed = candidate.trim();
+    if (!trimmed) {
+      return fallback;
+    }
+
+    const parsed = new URL(trimmed);
+    return ensureTrailingSlash(parsed.toString());
+  } catch (error) {
+    console.warn(
+      "Invalid client application URL provided, falling back to default domain",
+      error
+    );
+    return fallback;
+  }
+};
+
+const parseNumericQueryParam = (value: unknown) => {
+  if (typeof value !== "string") {
+    return null;
+  }
+
+  const parsed = Number.parseInt(value, 10);
+  return Number.isFinite(parsed) ? parsed : null;
+};
+
+const buildCheckoutRedirectUrl = (
+  baseUrl: string | null | undefined,
+  path: string,
+  params: URLSearchParams
+) => {
+  const target = new URL(path.replace(/^\/+/, ""), normaliseBaseUrl(baseUrl));
+  params.forEach((value, key) => {
+    target.searchParams.set(key, value);
+  });
+
+  return target.toString();
+};
+
 export default function () {
   const operations = {
     GET: [GET],
   };
 
   async function GET(req: Request, res: Response, next: NextFunction) {
-    const { success, session_id, stripeAccountId } = req.query;
+    const { success, canceled, session_id, stripeAccountId } = req.query;
     try {
+      if (canceled) {
+        const [clientId, artistId] = [
+          parseNumericQueryParam(req.query.clientId),
+          parseNumericQueryParam(req.query.artistId),
+        ];
```

Comment:
This feels like it would be much easier to read like this no?
```suggestion
        const clientId = parseNumericQueryParam(req.query.clientId);
        const artistId = parseNumericQueryParam(req.query.artistId);
```